### PR TITLE
client/http: do not retry the 4XX code returned by leader

### PR DESCRIPTION
**Pull request by JmPotato**
### What problem does this PR solve?

Issue Number: ref #7300.

### What is changed and how does it work?

```commit-message
When the PD leader returns a 404 status code, it no longer continues to retry but returns directly.
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/http/client.go [hunk 1]
```diff
@@ -147,16 +147,18 @@ func (ci *clientInner) requestWithRetry(
...
 	)
 	// Try to send the request to the PD leader first.
 	if leaderAddrIdx != -1 {
 		addr = pdAddrs[leaderAddrIdx]
-		err = ci.doRequest(ctx, addr, reqInfo, headerOpts...)
-		if err == nil {
-			return nil
+		statusCode, err = ci.doRequest(ctx, addr, reqInfo, headerOpts...)
+		// No need to retry if there is no error or the status code is 404 returned by the PD leader.
+		if err == nil || statusCode == http.StatusNotFound {
```
Comment (CabinfeverB) [inline client/http/client.go:160]: Also for StatusInternalServerError and StatusBadRequest?

Comment (rleungx) [reply client/http/client.go:160]: IMO, only the latter one is no need.

Comment (rleungx) [reply client/http/client.go:160]: BTW, we also have API which will return the 403 error.

Comment (CabinfeverB) [reply client/http/client.go:160]: Or we discuss what situations need retries?

Comment (JmPotato) [reply client/http/client.go]: We can first exclude some obvious scenarios that do not require retrying, and then supplement more later.

**Review by CabinfeverB — APPROVED** [review]

**Review by HuSharp — APPROVED** [review]

**Pull request merged**